src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/CanonicalEETypeNode.cs

```diff
@@ -0,0 +1,137 @@
+    /// they are used by the dynamic type loader when dynamically instantiating types at runtime.
+    /// The data that we emit on canonical type instantiations should just be the minimum that is needed by the template 
+    /// type loader. 
+    /// Similarly, the dependencies that we track for canonicl type instantiations are minimal, and are just the ones used
+    /// by the dynamic type loader
+    /// </summary>
+    internal sealed class CanonicalEETypeNode : EETypeNode
+    {
+        public CanonicalEETypeNode(NodeFactory factory, TypeDesc type) : base(factory, type)
+        {
+            Debug.Assert(!type.IsCanonicalDefinitionType(CanonicalFormKind.Any));
+            Debug.Assert(type.IsCanonicalSubtype(CanonicalFormKind.Any));
+            Debug.Assert(type == type.ConvertToCanonForm(CanonicalFormKind.Specific));
+
+            // TODO: needs a closer look when we enable USG
+            Debug.Assert(!type.IsCanonicalSubtype(CanonicalFormKind.Universal));
+        }
+
+        public override bool StaticDependenciesAreComputed => true;
+        public override bool IsShareable => IsTypeNodeShareable(_type);
+        public override bool HasConditionalStaticDependencies => false;
+
+        protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFactory factory)
+        {
+
+            DefType closestDefType = _type.GetClosestDefType();
+
+            DependencyList dependencyList = new DependencyList();
+
+            if (_type.RuntimeInterfaces.Length > 0)
+                dependencyList.Add(factory.InterfaceDispatchMap(_type), "Canonical interface dispatch map");
+
+            dependencyList.Add(factory.VTable(_type), "VTable");
+
+            // TODO: native layout dependencies (template type entries)
+
+            // TODO: other dependencies needed by the dynamic type loader?
+
+            return dependencyList;
+        }
+
+        protected override int GCDescSize
+        {
+            get
+            {
+                // TODO: handling of USG types
+                return GCDescEncoder.GetGCDescSize(_type);
+            }
+        }
+
+        protected override void OutputGCDesc(ref ObjectDataBuilder builder)
+        {
+            // TODO: handling of USG types
+            GCDescEncoder.EncodeGCDesc(ref builder, _type);
+        }
+
+        protected override void OutputVirtualSlots(NodeFactory factory, ref ObjectDataBuilder objData, TypeDesc implType, TypeDesc declType)
+        {
+            declType = declType.GetClosestDefType();
+
+            var baseType = declType.BaseType;
+            if (baseType != null)
+                OutputVirtualSlots(factory, ref objData, implType, baseType);
+
+            // Note: Canonical type instantiations always have a generic dictionary vtable slot, but it's empty
+            if (declType.HasGenericDictionarySlot())
+            {
+                Debug.Assert(declType.IsCanonicalSubtype(CanonicalFormKind.Any));
```

Comment:
Ok, looks like this code should really go on the EETypeNode. I don't see any derived EEType node that would override this and drastically change the implementation... I'll create virtual boolean flags instead to control the emission of such sections on the eetype (And derived eetypenodes can override these flags)

---
In reply to: [98532707](https://github.com/dotnet/corert/pull/2575#discussion_r98532707) [](ancestors = 98532707)